### PR TITLE
[libc][cmake] pass -mthumb for thumb triples

### PR DESCRIPTION
**Pull request by nickdesaulniers**
Allows cross compiling for 32b ARM in thumb mode via
`-DLIBC_TARGET_TRIPLE=thumbv7-linux-gnueabihf`. Strangely, setting that target
triple doesn't imply `-mthumb`, so set that when using a triple that starts
with "thumb".

While our continuous testing of 32b ARM of llvm-libc uses ARM mode (-marm), it might be nice
to simplify testing -mthumb so that we can set up a thumb build bot.

Link: #96566

### CODE REVIEW

**Comment by enh-google** [comment]
> Android only cares about building llvm-libc for 32b ARM in thumb
mode

citation needed? i really don't think we do... thumb might well be Android's _global default_ for arm32, but that's (a) not required and (b) not what we actually use for any of the syscall stuff in bionic...

**Comment by nickdesaulniers** [comment]
> > Android only cares about building llvm-libc for 32b ARM in thumb
> > mode
> 
> citation needed? i really don't think we do... thumb might well be Android's _global default_ for arm32, but that's (a) not required and (b) not what we actually use for any of the syscall stuff in bionic...

Let me rephrase.

While our continuous testing of 32b ARM of llvm-libc uses ARM mode (`-marm`), for Android support we also need to support Thumb mode (`-mthumb). ...

**Comment by enh-google** [comment]
> While our continuous testing of 32b ARM of llvm-libc uses ARM mode (`-marm`), for Android support we also need to support Thumb mode (`-mthumb). ...

why do you? bionic doesn't compile any of its syscall stubs as thumb. why should you?

(i'm not saying there aren't maybe microcontroller use cases or whatever, but (a) _Android_ doesn't care and (b) ILP32 will be completely dead to us soon anyway.)

**Comment by nickdesaulniers** [comment]
> why do you? bionic doesn't compile any of its syscall stubs as thumb. why should you?

When I last tried to build llvm-libc via the existing Android.bp files, I got an error that the syscall stubs didn't support `-mthumb`. So it seems that the default implicit compiler flags on android set `-mthumb`.

I don't personally care, but it seems that llvm-libc for Android is being built in thumb mode, so I thought it was important.  Should I be setting `-marm` for all of llvm-libc in the Android.bp file (for 32b arm), or just the syscalls?

> ILP32 will be completely dead to us soon anyway.)

So I've been hearing for years.

> Is that an issue in Clang?

Let me file a bug.

**Comment by nickdesaulniers** [comment]
> > Allows cross compiling for 32b ARM in thumb mode via `-DLIBC_TARGET_TRIPLE=thumbv7-linux-gnueabihf`. Strangely, setting that target triple doesn't imply `-mthumb`.
> 
> Is that an issue in Clang?

Filed #96566

**Comment by enh-google** [comment]
> > why do you? bionic doesn't compile any of its syscall stubs as thumb. why should you?
> 
> When I last tried to build llvm-libc via the existing Android.bp files, I got an error that the syscall stubs didn't support `-mthumb`. So it seems that the default implicit compiler flags on android set `-mthumb`.

yes, the global default is thumb, because -- by default, anyway -- why wouldn't we let the compiler use thumb instructions? you can override it with
```
            instruction_set: "arm",
```
in your .bp file. (we do this for all of libm, for example. except for the bits that come from external/arm-optimized-routines, though i assume that's just another "it's ILP32; nobody cares" oversight.)

> I don't personally care, but it seems that llvm-libc for Android is being built in thumb mode, so I thought it was important. Should I be setting `-marm` for all of llvm-libc in the Android.bp file (for 32b arm), or just the syscalls?

up to you. (given that the main _alleged_ reason for using the `arm` override is performance, i'd be happy to argue that it makes sense for all of libc.)

> > ILP32 will be completely dead to us soon anyway.)
> 
> So I've been hearing for years.

aye, it takes a very long time to ratchet these things down. the number of niches where you can still ship a 32-bit device is increasingly limited though.

**Comment by nickdesaulniers** [comment]
Ok, I can change Android's bp file to use `-marm` (perhaps worth benchmarking, but "eh 32b bit").

Orthogonal to that, I would like to:
1. be able to test `-mthumb` builds of llvm-libc
2. using the runtimes build

This cmake change enables simplicity of testing llvm-libc, so I think it stands on its own merit.  Unless @petrhosek knows of a better way to do this?
